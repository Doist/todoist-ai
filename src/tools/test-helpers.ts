import type { PersonalProject, Section, Task } from '@doist/todoist-api-typescript'

/**
 * Mapped task type matching the output of mapTask function.
 * Used for filter-based query test mocks.
 */
export type MappedTask = {
    id: string
    content: string
    description: string
    dueDate: string | undefined
    recurring: string | boolean
    priority: number
    projectId: string
    sectionId: string | null
    parentId: string | null
    labels: string[]
}

/**
 * Creates a mock Task with all required properties and sensible defaults.
 * Pass only the properties you want to override for your specific test.
 */
export function createMockTask(overrides: Partial<Task> = {}): Task {
    return {
        id: '8485093748',
        content: 'Test task content',
        description: '',
        completedAt: null,
        labels: [],
        childOrder: 1,
        priority: 1,
        projectId: '6cfCcrrCFg2xP94Q',
        sectionId: null,
        parentId: null,
        url: 'https://todoist.com/showTask?id=8485093748',
        // Use correct property names from Task schema
        noteCount: 0,
        addedByUid: '713437',
        addedAt: '2025-08-13T22:09:56.123456Z',
        deadline: null,
        responsibleUid: null,
        assignedByUid: null,
        isCollapsed: false,
        isDeleted: false,
        duration: null,
        checked: false,
        updatedAt: '2025-08-13T22:09:56.123456Z',
        due: null,
        dayOrder: 0,
        userId: '713437',
        ...overrides,
    }
}

/**
 * Creates a mock Section with all required properties and sensible defaults.
 * Pass only the properties you want to override for your specific test.
 */
export function createMockSection(overrides: Partial<Section> = {}): Section {
    return {
        id: 'section-123',
        projectId: '6cfCcrrCFg2xP94Q',
        sectionOrder: 1,
        userId: 'test-user',
        addedAt: '2024-01-01T00:00:00Z',
        updatedAt: '2024-01-01T00:00:00Z',
        archivedAt: null,
        isArchived: false,
        isDeleted: false,
        isCollapsed: false,
        name: 'Test Section',
        ...overrides,
    }
}

/**
 * Creates a mock PersonalProject with all required properties and sensible defaults.
 * Pass only the properties you want to override for your specific test.
 */
export function createMockProject(overrides: Partial<PersonalProject> = {}): PersonalProject {
    return {
        id: '6cfCcrrCFg2xP94Q',
        name: 'Test Project',
        color: 'charcoal',
        isFavorite: false,
        isShared: false,
        parentId: null,
        inboxProject: false,
        viewStyle: 'list',
        url: 'https://todoist.com/projects/6cfCcrrCFg2xP94Q',
        isDeleted: false,
        updatedAt: '2025-08-13T22:09:55.841800Z',
        createdAt: '2025-08-13T22:09:55.841785Z',
        childOrder: 1,
        defaultOrder: 0,
        description: '',
        isCollapsed: false,
        canAssignTasks: false,
        isFrozen: false,
        isArchived: false,
        ...overrides,
    }
}

/**
 * Creates a mock API response object with results and nextCursor.
 */
export function createMockApiResponse<T>(
    results: T[],
    nextCursor: string | null = null,
): {
    results: T[]
    nextCursor: string | null
} {
    return {
        results,
        nextCursor,
    }
}

/**
 * Creates a simplified mapped task (matches mapTask output) for filter-based query tests.
 */
export function createMappedTask(overrides: Partial<MappedTask> = {}): MappedTask {
    return {
        id: TEST_IDS.TASK_1,
        content: 'Test task content',
        description: '',
        dueDate: undefined,
        recurring: false,
        priority: 1,
        projectId: TEST_IDS.PROJECT_TEST,
        sectionId: null,
        parentId: null,
        labels: [],
        ...overrides,
    }
}

/**
 * Common error messages used across tests.
 */
export const TEST_ERRORS = {
    API_RATE_LIMIT: 'API Error: Rate limit exceeded',
    API_UNAUTHORIZED: 'API Error: Unauthorized',
    INVALID_CURSOR: 'Invalid cursor format',
    INVALID_FILTER: 'Invalid filter query',
} as const

/**
 * Creates multiple test cases for parameterized testing.
 */
export function createTestCases<T, E = unknown>(
    cases: Array<{ name: string; input: T; expected?: E }>,
) {
    return cases
}

/**
 * Common mock IDs used across tests for consistency.
 */
export const TEST_IDS = {
    TASK_1: '8485093748',
    TASK_2: '8485093749',
    TASK_3: '8485093750',
    PROJECT_INBOX: 'inbox-project-id',
    PROJECT_WORK: 'work-project-id',
    PROJECT_TEST: '6cfCcrrCFg2xP94Q',
    SECTION_1: 'section-123',
    SECTION_2: 'section-456',
    USER_ID: '713437',
} as const
