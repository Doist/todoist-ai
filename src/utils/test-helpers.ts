import type { PersonalProject, Section, Task } from '@doist/todoist-api-typescript'
import { getToolOutput } from '../mcp-helpers'

/**
 * Mapped task type matching the output of mapTask function.
 * Used for filter-based query test mocks.
 */
export type MappedTask = {
    id: string
    content: string
    description: string
    dueDate: string | undefined
    recurring: string | boolean
    priority: number
    projectId: string
    sectionId: string | null
    parentId: string | null
    labels: string[]
    duration: string | null
    responsibleUid: string | null
    assignedByUid: string | null
}

/**
 * Creates a mock Task with all required properties and sensible defaults.
 * Pass only the properties you want to override for your specific test.
 */
export function createMockTask(overrides: Partial<Task> = {}): Task {
    return {
        id: '8485093748',
        content: 'Test task content',
        description: '',
        completedAt: null,
        labels: [],
        childOrder: 1,
        priority: 1,
        projectId: '6cfCcrrCFg2xP94Q',
        sectionId: null,
        parentId: null,
        url: 'https://todoist.com/showTask?id=8485093748',
        // Use correct property names from Task schema
        noteCount: 0,
        addedByUid: '713437',
        addedAt: '2025-08-13T22:09:56.123456Z',
        deadline: null,
        responsibleUid: null,
        assignedByUid: null,
        isCollapsed: false,
        isDeleted: false,
        duration: null,
        checked: false,
        updatedAt: '2025-08-13T22:09:56.123456Z',
        due: null,
        dayOrder: 0,
        userId: '713437',
        ...overrides,
    }
}

/**
 * Creates a mock Section with all required properties and sensible defaults.
 * Pass only the properties you want to override for your specific test.
 */
export function createMockSection(overrides: Partial<Section> = {}): Section {
    return {
        id: 'section-123',
        projectId: '6cfCcrrCFg2xP94Q',
        sectionOrder: 1,
        userId: 'test-user',
        addedAt: '2024-01-01T00:00:00Z',
        updatedAt: '2024-01-01T00:00:00Z',
        archivedAt: null,
        isArchived: false,
        isDeleted: false,
        isCollapsed: false,
        name: 'Test Section',
        url: 'https://todoist.com/sections/section-123',
        ...overrides,
    }
}

/**
 * Creates a mock PersonalProject with all required properties and sensible defaults.
 * Pass only the properties you want to override for your specific test.
 */
export function createMockProject(overrides: Partial<PersonalProject> = {}): PersonalProject {
    return {
        id: '6cfCcrrCFg2xP94Q',
        name: 'Test Project',
        color: 'charcoal',
        isFavorite: false,
        isShared: false,
        parentId: null,
        inboxProject: false,
        viewStyle: 'list',
        url: 'https://todoist.com/projects/6cfCcrrCFg2xP94Q',
        isDeleted: false,
        updatedAt: '2025-08-13T22:09:55.841800Z',
        createdAt: '2025-08-13T22:09:55.841785Z',
        childOrder: 1,
        defaultOrder: 0,
        description: '',
        isCollapsed: false,
        canAssignTasks: false,
        isFrozen: false,
        isArchived: false,
        ...overrides,
    }
}

/**
 * Creates a mock API response object with results and nextCursor.
 */
export function createMockApiResponse<T>(
    results: T[],
    nextCursor: string | null = null,
): {
    results: T[]
    nextCursor: string | null
} {
    return {
        results,
        nextCursor,
    }
}

/**
 * Creates a simplified mapped task (matches mapTask output) for filter-based query tests.
 */
export function createMappedTask(overrides: Partial<MappedTask> = {}): MappedTask {
    return {
        id: TEST_IDS.TASK_1,
        content: 'Test task content',
        description: '',
        dueDate: undefined,
        recurring: false,
        priority: 1,
        projectId: TEST_IDS.PROJECT_TEST,
        sectionId: null,
        parentId: null,
        labels: [],
        duration: null,
        responsibleUid: null,
        assignedByUid: null,
        ...overrides,
    }
}

/**
 * Common error messages used across tests.
 */
export const TEST_ERRORS = {
    API_RATE_LIMIT: 'API Error: Rate limit exceeded',
    API_UNAUTHORIZED: 'API Error: Unauthorized',
    INVALID_CURSOR: 'Invalid cursor format',
    INVALID_FILTER: 'Invalid filter query',
} as const

/**
 * Creates multiple test cases for parameterized testing.
 */
export function createTestCases<T, E = unknown>(
    cases: Array<{ name: string; input: T; expected?: E }>,
) {
    return cases
}

/**
 * Extracts the text content from a tool output for snapshot testing.
 * This allows tests to match against just the text content while tools return structured output.
 */
export function extractTextContent(toolOutput: unknown): string {
    if (typeof toolOutput === 'string') {
        return toolOutput
    }

    if (typeof toolOutput === 'object' && toolOutput !== null && 'content' in toolOutput) {
        const output = toolOutput as { content: unknown }
        if (
            Array.isArray(output.content) &&
            output.content[0] &&
            typeof output.content[0] === 'object' &&
            output.content[0] !== null &&
            'type' in output.content[0] &&
            'text' in output.content[0] &&
            output.content[0].type === 'text'
        ) {
            return output.content[0].text as string
        }
    }

    throw new Error('Expected tool output to have text content')
}

/**
 * Extracts the structured content from a tool output for testing.
 * This handles both the new `structuredContent` field and legacy JSON-encoded content.
 */
export function extractStructuredContent(
    output: ReturnType<typeof getToolOutput>,
): Record<string, unknown> {
    // Check for new structuredContent field first
    if ('structuredContent' in output && typeof output.structuredContent === 'object') {
        return output.structuredContent as Record<string, unknown>
    }

    // Fall back to checking for JSON content in the content array
    if ('content' in output && Array.isArray(output.content)) {
        for (const item of output.content) {
            if (
                typeof item === 'object' &&
                item !== null &&
                'type' in item &&
                'text' in item &&
                item.type === 'text' &&
                item.mimeType === 'application/json'
            ) {
                return JSON.parse(item.text) as Record<string, unknown>
            }
        }
    }

    throw new Error('Expected tool output to have structured content')
}

/**
 * Common mock IDs used across tests for consistency.
 */
export const TEST_IDS = {
    TASK_1: '8485093748',
    TASK_2: '8485093749',
    TASK_3: '8485093750',
    PROJECT_INBOX: 'inbox-project-id',
    PROJECT_WORK: 'work-project-id',
    PROJECT_TEST: '6cfCcrrCFg2xP94Q',
    SECTION_1: 'section-123',
    SECTION_2: 'section-456',
    USER_ID: '713437',
} as const

/**
 * Fixed date for consistent test snapshots.
 * Use this instead of new Date() in tests to avoid snapshot drift.
 */
export const TODAY = '2025-08-17' as const
