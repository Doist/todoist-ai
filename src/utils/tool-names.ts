/**
 * Centralized tool names module
 *
 * This module provides a single source of truth for all tool names used throughout the codebase.
 * Each tool should import its own name from this module to avoid hardcoded strings.
 * This prevents outdated references when tool names change.
 */

export const TOOL_NAMES = {
    // Task management tools
    TASKS_ADD_MULTIPLE: 'tasks-add-multiple',
    TASKS_COMPLETE_MULTIPLE: 'tasks-complete-multiple',
    TASKS_UPDATE_MULTIPLE: 'tasks-update-multiple',
    TASKS_SEARCH: 'tasks-search',
    TASKS_LIST_BY_DATE: 'tasks-list-by-date',
    TASKS_LIST_COMPLETED: 'tasks-list-completed',
    TASKS_LIST_FOR_CONTAINER: 'tasks-list-for-container',

    // Project management tools
    PROJECTS_LIST: 'projects-list',
    PROJECTS_MANAGE: 'projects-manage',

    // Section management tools
    SECTIONS_SEARCH: 'sections-search',
    SECTIONS_MANAGE: 'sections-manage',

    // General tools
    OVERVIEW: 'overview',
    DELETE_ONE: 'delete-one',
} as const

// Type for all tool names
export type ToolName = (typeof TOOL_NAMES)[keyof typeof TOOL_NAMES]
