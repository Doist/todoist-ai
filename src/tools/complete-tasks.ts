import { z } from 'zod'
import { getToolOutput } from '../mcp-helpers.js'
import type { TodoistTool } from '../todoist-tool.js'
import { summarizeBatch } from '../utils/response-builders.js'
import { ToolNames } from '../utils/tool-names.js'

const ArgsSchema = {
    ids: z.array(z.string().min(1)).min(1).describe('The IDs of the tasks to complete.'),
}

const completeTasks = {
    name: ToolNames.COMPLETE_TASKS,
    description: 'Complete one or more tasks by their IDs.',
    parameters: ArgsSchema,
    async execute(args, client) {
        const completed: string[] = []
        const failures: Array<{ item: string; error: string; code?: string }> = []

        for (const id of args.ids) {
            try {
                await client.closeTask(id)
                completed.push(id)
            } catch (error) {
                const errorMessage = error instanceof Error ? error.message : 'Unknown error'
                failures.push({
                    item: id,
                    error: errorMessage,
                })
            }
        }

        const textContent = generateTextContent({
            completed,
            failures,
            args,
        })

        return getToolOutput({
            textContent,
            structuredContent: {
                completed,
                failures,
                totalRequested: args.ids.length,
                successCount: completed.length,
                failureCount: failures.length,
            },
        })
    },
} satisfies TodoistTool<typeof ArgsSchema>

function generateNextSteps(completed: number, failures: number): string[] {
    if (completed > 0) {
        const moveResult =
            failures === 0
                ? "Use find-tasks-by-date('today') to tackle remaining overdue items."
                : 'Review failed completions and retry if needed.'
        return [moveResult]
    }

    if (failures > 0) {
        return ['Check task IDs and permissions, then retry.']
    }

    return ['No tasks were completed.']
}

function generateTextContent({
    completed,
    failures,
    args,
}: {
    completed: string[]
    failures: Array<{ item: string; error: string; code?: string }>
    args: z.infer<z.ZodObject<typeof ArgsSchema>>
}) {
    const nextSteps = generateNextSteps(completed.length, failures.length)
    return summarizeBatch({
        action: 'Completed tasks',
        success: completed.length,
        total: args.ids.length,
        successItems: completed,
        failures,
        nextSteps,
    })
}

export { completeTasks }
