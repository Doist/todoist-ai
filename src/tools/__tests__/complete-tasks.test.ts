import type { TodoistApi } from '@doist/todoist-api-typescript'
import { jest } from '@jest/globals'
import { extractTextContent } from '../../utils/test-helpers.js'
import { ToolNames } from '../../utils/tool-names.js'
import { completeTasks } from '../complete-tasks.js'

// Mock the Todoist API
const mockTodoistApi = {
    closeTask: jest.fn(),
} as unknown as jest.Mocked<TodoistApi>

const { COMPLETE_TASKS, GET_OVERVIEW } = ToolNames

describe(`${COMPLETE_TASKS} tool`, () => {
    beforeEach(() => {
        jest.clearAllMocks()
    })

    describe('completing multiple tasks', () => {
        it('should complete all tasks successfully', async () => {
            mockTodoistApi.closeTask.mockResolvedValue(true)

            const result = await completeTasks.execute(
                { ids: ['task-1', 'task-2', 'task-3'] },
                mockTodoistApi,
            )

            // Verify API was called for each task
            expect(mockTodoistApi.closeTask).toHaveBeenCalledTimes(3)
            expect(mockTodoistApi.closeTask).toHaveBeenNthCalledWith(1, 'task-1')
            expect(mockTodoistApi.closeTask).toHaveBeenNthCalledWith(2, 'task-2')
            expect(mockTodoistApi.closeTask).toHaveBeenNthCalledWith(3, 'task-3')

            // Verify result is a concise summary
            expect(extractTextContent(result)).toMatchSnapshot()

            // Verify structured content
            const { structuredContent } = result
            expect(structuredContent).toEqual(
                expect.objectContaining({
                    completed: ['task-1', 'task-2', 'task-3'],
                    failures: [],
                    totalRequested: 3,
                    successCount: 3,
                    failureCount: 0,
                }),
            )
        })

        it('should complete single task', async () => {
            mockTodoistApi.closeTask.mockResolvedValue(true)

            const result = await completeTasks.execute({ ids: ['8485093748'] }, mockTodoistApi)

            expect(mockTodoistApi.closeTask).toHaveBeenCalledTimes(1)
            expect(mockTodoistApi.closeTask).toHaveBeenCalledWith('8485093748')

            expect(extractTextContent(result)).toMatchSnapshot()

            // Verify structured content
            const { structuredContent } = result
            expect(structuredContent).toEqual(
                expect.objectContaining({
                    completed: ['8485093748'],
                    failures: [],
                    totalRequested: 1,
                    successCount: 1,
                    failureCount: 0,
                }),
            )
        })

        it('should handle partial failures gracefully', async () => {
            // Mock first and third tasks to succeed, second to fail
            mockTodoistApi.closeTask
                .mockResolvedValueOnce(true) // task-1 succeeds
                .mockRejectedValueOnce(new Error('Task not found')) // task-2 fails
                .mockResolvedValueOnce(true) // task-3 succeeds

            const result = await completeTasks.execute(
                { ids: ['task-1', 'task-2', 'task-3'] },
                mockTodoistApi,
            )

            // Verify API was called for all tasks despite failure
            expect(mockTodoistApi.closeTask).toHaveBeenCalledTimes(3)
            expect(mockTodoistApi.closeTask).toHaveBeenNthCalledWith(1, 'task-1')
            expect(mockTodoistApi.closeTask).toHaveBeenNthCalledWith(2, 'task-2')
            expect(mockTodoistApi.closeTask).toHaveBeenNthCalledWith(3, 'task-3')

            // Verify only successful completions are reported
            expect(extractTextContent(result)).toMatchSnapshot()

            // Verify structured content with partial failures
            const { structuredContent } = result
            expect(structuredContent).toEqual(
                expect.objectContaining({
                    completed: ['task-1', 'task-3'],
                    failures: [
                        expect.objectContaining({
                            item: 'task-2',
                            error: 'Task not found',
                        }),
                    ],
                    totalRequested: 3,
                    successCount: 2,
                    failureCount: 1,
                }),
            )
        })

        it('should handle all tasks failing', async () => {
            const apiError = new Error('API Error: Network timeout')
            mockTodoistApi.closeTask.mockRejectedValue(apiError)

            const result = await completeTasks.execute(
                { ids: ['task-1', 'task-2'] },
                mockTodoistApi,
            )

            // Verify API was attempted for all tasks
            expect(mockTodoistApi.closeTask).toHaveBeenCalledTimes(2)

            // Verify no tasks were completed but still returns success
            expect(extractTextContent(result)).toMatchSnapshot()
        })

        it('should continue processing remaining tasks after failures', async () => {
            // Mock various failure scenarios
            mockTodoistApi.closeTask
                .mockRejectedValueOnce(new Error('Task already completed'))
                .mockRejectedValueOnce(new Error('Task not found'))
                .mockResolvedValueOnce(true) // task-3 succeeds
                .mockRejectedValueOnce(new Error('Permission denied'))
                .mockResolvedValueOnce(true) // task-5 succeeds

            const result = await completeTasks.execute(
                { ids: ['task-1', 'task-2', 'task-3', 'task-4', 'task-5'] },
                mockTodoistApi,
            )

            expect(mockTodoistApi.closeTask).toHaveBeenCalledTimes(5)

            // Only tasks 3 and 5 should be in completed list
            expect(extractTextContent(result)).toMatchSnapshot()
        })

        it('should handle different types of API errors', async () => {
            mockTodoistApi.closeTask
                .mockRejectedValueOnce(new Error('Task not found'))
                .mockRejectedValueOnce(new Error('Task already completed'))
                .mockRejectedValueOnce(new Error('Permission denied'))
                .mockRejectedValueOnce(new Error('Rate limit exceeded'))

            const result = await completeTasks.execute(
                { ids: ['not-found', 'already-done', 'no-permission', 'rate-limited'] },
                mockTodoistApi,
            )

            expect(mockTodoistApi.closeTask).toHaveBeenCalledTimes(4)

            // All should fail, but the tool should handle it gracefully
            expect(extractTextContent(result)).toMatchSnapshot()
        })
    })

    describe('mixed success and failure scenarios', () => {
        it('should handle realistic mixed scenario', async () => {
            // Simulate a realistic scenario with some tasks completing and others failing
            mockTodoistApi.closeTask
                .mockResolvedValueOnce(true) // regular task completion
                .mockResolvedValueOnce(true) // another successful completion
                .mockRejectedValueOnce(new Error('Task already completed')) // duplicate completion
                .mockResolvedValueOnce(true) // successful completion
                .mockRejectedValueOnce(new Error('Task not found')) // deleted task

            const result = await completeTasks.execute(
                {
                    ids: [
                        '8485093748', // regular task
                        '8485093749', // regular task
                        '8485093750', // already completed
                        '8485093751', // regular task
                        '8485093752', // deleted task
                    ],
                },
                mockTodoistApi,
            )

            expect(mockTodoistApi.closeTask).toHaveBeenCalledTimes(5)

            expect(extractTextContent(result)).toMatchSnapshot()
        })
    })

    describe('next steps logic validation', () => {
        it('should suggest overdue tasks when all tasks complete successfully', async () => {
            mockTodoistApi.closeTask.mockResolvedValue(true)

            const result = await completeTasks.execute(
                { ids: ['task-1', 'task-2'] },
                mockTodoistApi,
            )

            const textContent = extractTextContent(result)
            expect(textContent).toMatchSnapshot()
            expect(textContent).toContain("Use find-tasks-by-date('today')")
        })

        it('should suggest reviewing failures when mixed results', async () => {
            mockTodoistApi.closeTask
                .mockResolvedValueOnce(true)
                .mockRejectedValueOnce(new Error('Task not found'))

            const result = await completeTasks.execute(
                { ids: ['task-1', 'task-2'] },
                mockTodoistApi,
            )

            const textContent = extractTextContent(result)
            expect(textContent).toMatchSnapshot()
            expect(textContent).toContain('Review failed completions and retry if needed')
        })

        it('should suggest checking IDs when all tasks fail', async () => {
            mockTodoistApi.closeTask.mockRejectedValue(new Error('Task not found'))

            const result = await completeTasks.execute(
                { ids: ['bad-id-1', 'bad-id-2'] },
                mockTodoistApi,
            )

            const textContent = extractTextContent(result)
            expect(textContent).toMatchSnapshot()
            expect(textContent).toContain('Check task IDs and permissions, then retry')
            expect(textContent).not.toContain(`Use ${GET_OVERVIEW} tool`) // Should only show retry message
        })
    })

    describe('error message truncation', () => {
        it('should truncate failure messages after 3 errors', async () => {
            mockTodoistApi.closeTask
                .mockRejectedValueOnce(new Error('Error 1'))
                .mockRejectedValueOnce(new Error('Error 2'))
                .mockRejectedValueOnce(new Error('Error 3'))
                .mockRejectedValueOnce(new Error('Error 4'))
                .mockRejectedValueOnce(new Error('Error 5'))

            const result = await completeTasks.execute(
                { ids: ['task-1', 'task-2', 'task-3', 'task-4', 'task-5'] },
                mockTodoistApi,
            )

            const textContent = extractTextContent(result)
            expect(textContent).toMatchSnapshot()
            expect(textContent).toContain('+2 more') // 5 total failures, showing first 3, so +2 more
            expect(textContent).not.toContain('Error 4') // Should not show 4th error
            expect(textContent).not.toContain('Error 5') // Should not show 5th error
        })

        it('should not show truncation message for exactly 3 errors', async () => {
            mockTodoistApi.closeTask
                .mockRejectedValueOnce(new Error('Error 1'))
                .mockRejectedValueOnce(new Error('Error 2'))
                .mockRejectedValueOnce(new Error('Error 3'))

            const result = await completeTasks.execute(
                { ids: ['task-1', 'task-2', 'task-3'] },
                mockTodoistApi,
            )

            const textContent = extractTextContent(result)
            expect(textContent).toMatchSnapshot()
            expect(textContent).not.toContain('more') // Should not show truncation
        })
    })

    describe('edge cases', () => {
        it('should handle empty task completion (minimum one task required by schema)', async () => {
            // Note: This test documents that the schema requires at least one task,
            // so this scenario shouldn't occur in practice due to validation
            mockTodoistApi.closeTask.mockResolvedValue(true)

            const result = await completeTasks.execute({ ids: ['single-task'] }, mockTodoistApi)

            expect(extractTextContent(result)).toMatchSnapshot()
        })

        it('should handle tasks with special ID formats', async () => {
            mockTodoistApi.closeTask.mockResolvedValue(true)

            const result = await completeTasks.execute(
                { ids: ['proj_123_task_456', 'task-with-dashes', '1234567890'] },
                mockTodoistApi,
            )

            expect(mockTodoistApi.closeTask).toHaveBeenCalledTimes(3)
            expect(mockTodoistApi.closeTask).toHaveBeenCalledWith('proj_123_task_456')
            expect(mockTodoistApi.closeTask).toHaveBeenCalledWith('task-with-dashes')
            expect(mockTodoistApi.closeTask).toHaveBeenCalledWith('1234567890')

            expect(extractTextContent(result)).toMatchSnapshot()
        })
    })
})
