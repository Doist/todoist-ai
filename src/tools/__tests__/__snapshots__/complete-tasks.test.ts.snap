// Jest Snapshot v1, https://jestjs.io/docs/snapshot-testing

exports[`complete-tasks tool completing multiple tasks should complete all tasks successfully 1`] = `
"Completed tasks: 3/3 successful.
Completed:
    task-1
    task-2
    task-3.
Next:
- Use find-tasks-by-date('today') to tackle remaining overdue items."
`;

exports[`complete-tasks tool completing multiple tasks should complete single task 1`] = `
"Completed tasks: 1/1 successful.
Completed:
    8485093748.
Next:
- Use find-tasks-by-date('today') to tackle remaining overdue items."
`;

exports[`complete-tasks tool completing multiple tasks should continue processing remaining tasks after failures 1`] = `
"Completed tasks: 2/5 successful.
Completed:
    task-3
    task-5.
Failed (3):
    task-1 (Error: Task already completed)
    task-2 (Error: Task not found)
    task-4 (Error: Permission denied).
Next:
- Review failed completions and retry if needed."
`;

exports[`complete-tasks tool completing multiple tasks should handle all tasks failing 1`] = `
"Completed tasks: 0/2 successful.
Failed (2):
    task-1 (Error: API Error: Network timeout)
    task-2 (Error: API Error: Network timeout).
Next:
- Check task IDs and permissions, then retry."
`;

exports[`complete-tasks tool completing multiple tasks should handle different types of API errors 1`] = `
"Completed tasks: 0/4 successful.
Failed (4):
    not-found (Error: Task not found)
    already-done (Error: Task already completed)
    no-permission (Error: Permission denied), +1 more.
Next:
- Check task IDs and permissions, then retry."
`;

exports[`complete-tasks tool completing multiple tasks should handle partial failures gracefully 1`] = `
"Completed tasks: 2/3 successful.
Completed:
    task-1
    task-3.
Failed (1):
    task-2 (Error: Task not found).
Next:
- Review failed completions and retry if needed."
`;

exports[`complete-tasks tool edge cases should handle empty task completion (minimum one task required by schema) 1`] = `
"Completed tasks: 1/1 successful.
Completed:
    single-task.
Next:
- Use find-tasks-by-date('today') to tackle remaining overdue items."
`;

exports[`complete-tasks tool edge cases should handle tasks with special ID formats 1`] = `
"Completed tasks: 3/3 successful.
Completed:
    proj_123_task_456
    task-with-dashes
    1234567890.
Next:
- Use find-tasks-by-date('today') to tackle remaining overdue items."
`;

exports[`complete-tasks tool error message truncation should not show truncation message for exactly 3 errors 1`] = `
"Completed tasks: 0/3 successful.
Failed (3):
    task-1 (Error: Error 1)
    task-2 (Error: Error 2)
    task-3 (Error: Error 3).
Next:
- Check task IDs and permissions, then retry."
`;

exports[`complete-tasks tool error message truncation should truncate failure messages after 3 errors 1`] = `
"Completed tasks: 0/5 successful.
Failed (5):
    task-1 (Error: Error 1)
    task-2 (Error: Error 2)
    task-3 (Error: Error 3), +2 more.
Next:
- Check task IDs and permissions, then retry."
`;

exports[`complete-tasks tool mixed success and failure scenarios should handle realistic mixed scenario 1`] = `
"Completed tasks: 3/5 successful.
Completed:
    8485093748
    8485093749
    8485093751.
Failed (2):
    8485093750 (Error: Task already completed)
    8485093752 (Error: Task not found).
Next:
- Review failed completions and retry if needed."
`;

exports[`complete-tasks tool next steps logic validation should suggest checking IDs when all tasks fail 1`] = `
"Completed tasks: 0/2 successful.
Failed (2):
    bad-id-1 (Error: Task not found)
    bad-id-2 (Error: Task not found).
Next:
- Check task IDs and permissions, then retry."
`;

exports[`complete-tasks tool next steps logic validation should suggest overdue tasks when all tasks complete successfully 1`] = `
"Completed tasks: 2/2 successful.
Completed:
    task-1
    task-2.
Next:
- Use find-tasks-by-date('today') to tackle remaining overdue items."
`;

exports[`complete-tasks tool next steps logic validation should suggest reviewing failures when mixed results 1`] = `
"Completed tasks: 1/2 successful.
Completed:
    task-1.
Failed (1):
    task-2 (Error: Task not found).
Next:
- Review failed completions and retry if needed."
`;
